### DISCUSSION TITLE
Skip string normalization in formatter

### DISCUSSION THREAD

**Discussion by vanchaxy** [astral-sh/ruff#7305] [Q&A] [answered]
We are trying to test the Ruff formatter as a replacement for Black. Does Ruff have an alternative to the "[--skip-string-normalization](https://black.readthedocs.io/en/stable/the_black_code_style/current_style.html#strings)" option from Black?

**Comment by MichaReiser** [comment, marked as answer]
Hey

No, ruff doesn't support `--skip-string-normalization` although it shouldn't be too hard to implement. What's your reason for using it? Is it because you prefer single quotes over double quotes or do you use a mix of both?

**Reply by vanchaxy** [reply]
Thanks for your answer! We prefer single quotes. Currently, we use `--skip-string-normalization` + ruff autofix to force single quotes everywhere.

Are there any plans/estimates of when the ruff formatter will support a configurable quote style? Or, maybe, an issue to subscribe to?

**Reply by MichaReiser** [reply]
Great. Ruff's formatter already supports single or double quotes, it's just that the option isn't yet wired to the configuration because there are some unanswered questions about how the linter/formatter configuration should co-exist. 

There's no definite timeline but we plan to add configuration support as part of our [beta release](https://github.com/astral-sh/ruff/milestone/5). Expect it to take a couple of weeks, maybe even more, depending on the user feedback.

**Reply by alkatar21** [reply]
Hey, in all projects I'm involved with, the option is also used, however both styles are actually used. Maybe this should be changed in the long run, but currently mainly single quotes are used, except for f/r-strings.

**Reply by MichaReiser** [reply]
Thank you @alkatar21. What's the reason for not migrating the project to use a unified quote style? Is it because you want to avoid large formatter PRs and them showing in blame views (did you know that you can exclude commits from [blame view](https://docs.github.com/en/repositories/working-with-files/using-files/viewing-a-file#ignore-commits-in-the-blame-view)?), because you prefer single quotes and black doesn't support them, or is there another reason? 

I'm asking because the main purpose of a formatter is to enforce consistency, and this option defeats this purpose. Furthermore, Black only added the option for backward compatibility, which is why I would want to avoid adding the option, but only if there aren't any use cases that we're overlooking.

**Reply by alkatar21** [reply]
I can't test it right now (how it would behave in blue for example), but one reason I think was that strings in f-strings should have single quotes(even if they are added later, without having to change the outer quotes then). So it should all be consistent, though unfortunately not ensured by the formatter.
```python
f"Ping {pong}"
f"Ping {pong['pong']}!"
'Everything else'
```

So basically every simple string should have single quotes. Large formatter PRs should not be necessary. It's probably more a question of behavior for strings in f-strings? Is the quote style considered from inside to outside or vice versa? As soon as the option for single quotes is available, I can give this a try.

**Reply by MichaReiser** [reply]
I don't think what you want would work because the formatter normalizes the f-string quotes to single quotes by default:

```python
f'Ping {pong}'
f"Ping {pong['pong']}!"
'Everything else'
```

You can see how it preserves the double quotes for the second example but changes the quotes to single quotes for the first example. 

Luckily, this will be fixed by Python 3.12 that allows strings inside of f-strings but you may not be able to use it right away, depending on which Python version you're targeting.

**Reply by tdulcet** [reply]
For my projects, the quote normalization would be OK, as this should be already fixed by the `Q` rules in the linter. However, the prefix normalization would be very problematic, as it breaks backwards compatibility with some older versions of Pythons that we still need to support for a couple of our projects. Specifically, the issue is the change from `br"..."` to `rb"..."` for binary regular expressions, which causes a `SyntaxError`. Without some type of replacement for this `--skip-string-normalization` option, we would not be able to switch to Ruff for formatting.

**Reply by MichaReiser** [reply]
@tdulcet do you know which Python versions require `br` over `rb`?

**Reply by tdulcet** [reply]
Not versions that are technically supported by Ruff. Specifically, Python 3.1 and older, including 2.7 which we still need to support unfortunately...

**Reply by MichaReiser** [reply]
Thanks @tdulcet. I included your input in the issue discussion what target versions ruff's formatter should support #7234

**Reply by DDuarte** [reply]
Adding another use case for this very much helpful option - incremental formatting

For years I have a project that had no style guide so there's a mix of single and double quotes everywhere. To start converging code to a single code style, a formatter was added to Pull Requests (+ precommit for local) that will format changed files in the PR. Without `--skip-string-normalization` most of the diffs produced are quote changes, polluting the diff. With the option, new code will be written with double quotes and _eventually_ everything will be formatted the same way. I guess I could just run the formatter on all code but that has other cons... Cheers

**Reply by HappyCthulhu** [reply]
Mb i will seem like a code monkey, but i dont rly find single/double quote issue important for some of my projects. I mean, some coders used do type single quotes, some - double. I wanna to left freedom of choice, bc, as i texted before - that isnt such a big deal for me.

**Reply by edreamleo** [reply]
Hi, I am the author of [Leo](https://leo-editor.github.io/leo-editor/) a large open-source project.  `--skip-string-normalization` is essential for Leo:

- To retain Leo's code style: Leo uses single and double quotes for distinct purposes.
- To eliminate large diffs in hundreds of files. This will allow incremental formatting, as mentioned in another response.

**Reply by MichaReiser** [reply]
Could you take a look at https://github.com/astral-sh/ruff/pull/8822 and leave a comment if the new introduced option isn't sufficient for your use case of `--skip-string-normalization`. Thanks.

**Reply by DDuarte** [reply]
`quote-style: preserve` looks perfect
